### DISCUSSION TITLE
Use of TransformRigidityPenalty

### DISCUSSION THREAD

**Discussion by cagatayalici** [SuperElastix/elastix#981] [Q&A]
Hi, 

I use 3 different parameter maps with a multi-resolution approach to register pretreatment and posttreatment lung CTs to evaluate tumor treatment ( Rigid -> Affine -> B-Spline). I want a specific region defined by a segmentation not to be deformed much during nonlinear registration. 

For that, I introduced TransformRigidityPenalty where I used Elastix manual and [ "A rigidity penalty term for nonrigid registration."](https://pubmed.ncbi.nlm.nih.gov/18072476/) as reference. However, I am getting weird results for the registered image as can be seen below: 

![registered_image](https://github.com/SuperElastix/elastix/assets/62696511/f20a6d9e-a84d-4e15-a0ea-5f52c007dbe1)

Below you can see the red region in 3D which I want it not to be deformed:

<img src="https://github.com/SuperElastix/elastix/assets/62696511/27bc0ecc-909b-4041-8f75-62f40fa7f991" width="300">

I only used TransformRigidityPenalty in 3rd parameter map which is B-Spline. Beloware the parameters I used:

```
// **************** Main Components **************************
(FixedInternalImagePixelType "float")
(MovingInternalImagePixelType "float")
(FixedImageDimension 3)
(MovingImageDimension 3)
(Registration "MultiMetricMultiResolutionRegistration")
(Interpolator "BSplineInterpolator")
(ResampleInterpolator "FinalBSplineInterpolator")
(Resampler "DefaultResampler")
(FixedImagePyramid "FixedSmoothingImagePyramid")
(MovingImagePyramid "MovingSmoothingImagePyramid")

// ***************** Transformation **************************
(Transform "BSplineTransform")
(HowToCombineTransforms "Compose")

// ******************* Similarity measure *********************
(Metric "AdvancedMattesMutualInformation" "TransformRigidityPenalty")
(UseRelativeWeights "false")
(Metric0Weight 1.0)
(Metric0Use "true")
(Metric1Weight 1.0)
(Metric1Use "true")
(LinearityConditionWeight 100.0)
(OrthonormalityConditionWeight 1.0)
(PropernessConditionWeight 2.0)
(NumberOfHistogramBins 64)

// ******************** Multiresolution **********************
(NumberOfResolutions 4)
(ImagePyramidSchedule 4 4 4 4 4 4 2 2 2 1 1 1)
(FinalGridSpacingInVoxels 8.0 8.0 8.0)
(GridSpacingSchedule 6.0 6.0 6.0 4.0 4.0 4.0 2.0 2.0 2.0 1.0 1.0 1.0)

// ******************* Optimizer ****************************
(Optimizer "AdaptiveStochasticGradientDescent")
(AutomaticParameterEstimation "true")
(UseAdaptiveStepSizes "true")
(MaximumNumberOfIterations 300 300 300 1200)

// **************** Image sampling **********************
(NumberOfSpatialSamples 5000)
(NewSamplesEveryIteration "true")
(ImageSampler "RandomCoordinate")

// ************* Interpolation and Resampling ****************
(BSplineInterpolationOrder 3)
(FinalBSplineInterpolationOrder 3)
(DefaultPixelValue -1024)
```

And I set  the rigidity index of voxels in the moving image via:  `parameter_object.SetParameter(2, "MovingRigidityImageName", "moving_rigidity_mask.nrrd"))`

With iterations TransformRigidityPenalty gets less however, it is still too high: 
<img src="https://github.com/SuperElastix/elastix/assets/62696511/d7ccedcb-b8a4-452f-a437-1ae000627fe9" width="300">

What could be the reason for black regions formed in the registered image?

Thank you!

**Comment by costanzasimoncini** [comment]
Hello, did you understand the reason of this behaviour? I have the same problem but only on Mac OS. I tried the exact same images, same parameters, same Elastix version (5.0.1) on windows and everything works fine... really strange.

**Reply by cagatayalici** [reply]
Hi,

Unfortunately, I don't know the root cause of it.  

I realized that at some point during iteration, the penalty metric becomes so large that it is taken as infinite, and then something happens in the algorithm (probably creating those distorted regions), and the penalty becomes zero afterward. 

I found another parameter set that seems to be working. You may find them on the [itk forum](https://discourse.itk.org/t/use-of-transformrigiditypenalty-fails/6273/2).

But your case sounds strange.
